File: src/components/wifi/efWifiModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo'

type EfWifiRequest = {
  ssid: string
  identity: string
  password: string
  anonymousIdentity?: string
  domainSuffixMatch?: string
}

declare class EfWifiNativeModule extends NativeModule<{}> {
  addEnterpriseNetwork(request: EfWifiRequest): Promise<void>
}

const efWifiModule = requireOptionalNativeModule<EfWifiNativeModule>('EfWifi')

// Native module (see EfWifiModule.kt) launches the Android system "add network" dialog on
// API 30+ and throws EfWifiException / MissingActivity on failure rather than returning a code.
export async function addEnterpriseNetwork(
  ssid: string,
  identity: string,
  password: string,
  anonymousIdentity: string,
  domainSuffixMatch: string
): Promise<void> {
  if (!efWifiModule)
    throw new Error('EfWifi native module is not available in this build')
  await efWifiModule.addEnterpriseNetwork({
    ssid,
    identity,
    password,
    anonymousIdentity,
    domainSuffixMatch,
  })
}

File: modules/ef-wifi/android/src/main/java/expo/modules/efwifi/EfWifiModule.kt
package expo.modules.efwifi

import android.content.Intent
import android.net.wifi.WifiEnterpriseConfig
import android.net.wifi.WifiNetworkSuggestion
import android.os.Build
import android.provider.Settings
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.InputStream
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate

class EfWifiException(message: String) : CodedException(message)

class EfWifiModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("EfWifi")

    AsyncFunction("addEnterpriseNetwork") { request: Map<String, Any?> ->
      val activity = appContext.currentActivity ?: throw Exceptions.MissingActivity()

      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
        throw EfWifiException("Adding a network requires Android 11 (API 30) or newer.")
      }

      val ssid = request["ssid"] as? String ?: throw EfWifiException("Missing ssid")
      val identity = request["identity"] as? String ?: throw EfWifiException("Missing identity")
      val password = request["password"] as? String ?: throw EfWifiException("Missing password")
      val anonymousIdentity = request["anonymousIdentity"] as? String ?: "anonymous"
      val domainSuffixMatch = request["domainSuffixMatch"] as? String ?: ""

      // SECURITY: load bundled CA + set domain suffix match so the RADIUS server cert is validated.
      val context = activity.applicationContext
      val cf = CertificateFactory.getInstance("X.509")
      val caInput: InputStream = context.resources.openRawResource(R.raw.efwifi_ca)
      val caCert = caInput.use { cf.generateCertificate(it) as X509Certificate }

      val enterprise = WifiEnterpriseConfig().apply {
        eapMethod = WifiEnterpriseConfig.Eap.TTLS
        phase2Method = WifiEnterpriseConfig.Phase2.PAP
        this.identity = identity
        this.anonymousIdentity = anonymousIdentity
        this.password = password
        caCertificate = caCert
        if (domainSuffixMatch.isNotEmpty()) this.domainSuffixMatch = domainSuffixMatch
      }

      val suggestion = WifiNetworkSuggestion.Builder()
        .setSsid(ssid)
        .setWpa2EnterpriseConfig(enterprise)
        .build()

      val intent = Intent(Settings.ACTION_WIFI_ADD_NETWORKS).apply {
        putParcelableArrayListExtra(Settings.EXTRA_WIFI_NETWORK_LIST, arrayListOf(suggestion))
      }
      activity.startActivity(intent)
    }
  }
}
